// zmp-openapi-nodejs/build/main/dto/base-client.dto.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseClient = void 0;
const axios_1 = __importStar(require("axios"));
const form_data_1 = __importDefault(require("form-data"));
const qs_1 = __importDefault(require("qs"));
const constant_1 = require("../config/constant");
class BaseClient {
    constructor(version, apiKeyHeaderName, apiKey, identityHeaderName, identity, proxy) {
        this.isUseProxy = false;
        this.domain = constant_1.Constant.DOMAIN_PROD;
        this.sdkVersion = '2.1.0';
        this.sdkName = 'Node.js';
        this.validateInit = () => {
            if (!this.apiKey ||
                !this.identity ||
                !this.apiKeyHeaderName ||
                !this.identityHeaderName) {
                throw new Error('Invalid init value');
            }
            if (this.isUseProxy && (!this.proxy.host || !this.proxy.port)) {
                throw new Error('Invalid proxy value');
            }
        };
        this.getVersionsMiniApp = (appSlice) => {
            return this.doGet(`${this.buildRequestURI(appSlice.miniAppId, constant_1.Constant.VERSIONS)}`, Object.assign({}, appSlice), {
                mapAxiosError(error) {
                    if (error.response.status === 400) {
                        return 'Invalid Mini App ID';
                    }
                    return error.message;
                },
            });
        };
        this.createMiniApp = (appInfo) => {
            return this.doPost(`${this.buildRequestURI()}`, Object.assign({}, appInfo));
        };
        this.deployMiniApp = (deployApp) => {
            const formData = new form_data_1.default();
            formData.append('file', deployApp.file, {
                contentType: 'application/zip',
            });
            return this.doPost(`${this.buildRequestURI(deployApp.miniAppId, constant_1.Constant.UPLOAD)}?${qs_1.default.stringify({
                name: deployApp.name,
                description: deployApp.description,
            })}`, formData);
        };
        this.requestPublishMiniApp = (requestPublishApp) => {
            return this.doPost(`${this.buildRequestURI(requestPublishApp.miniAppId, constant_1.Constant.REQUEST_PUBLISH)}`, Object.assign({}, requestPublishApp));
        };
        this.publishMiniApp = (publishApp) => {
            return this.doPost(`${this.buildRequestURI(publishApp.miniAppId, constant_1.Constant.PUBLISH)}`, Object.assign({}, publishApp));
        };
        this.getStats = async (statsRequest) => {
            const response = await this.doGet(`${this.buildRequestURI(statsRequest.miniAppId, constant_1.Constant.STATS)}`, Object.assign({}, statsRequest));
            const { error, message } = response, data = __rest(response, ["error", "message"]);
            return {
                error,
                message,
                data,
            };
        };
        this.listPaymentChannels = async (request) => {
            return this.doGet(`${this.buildRequestURI(request.miniAppId, constant_1.Constant.PAYMENT_CHANNELS)}`, {});
        };
        this.convertObjectToFormData = (obj) => {
            const formData = new form_data_1.default();
            for (const key in obj) {
                if (obj.hasOwnProperty(key)) {
                    const value = obj[key];
                    if (Array.isArray(value)) {
                        for (let i = 0; i < value.length; i++) {
                            formData.append(`${key}[${i}]`, value[i]);
                        }
                    }
                    else if (typeof obj[key] === 'boolean') {
                        formData.append(key, obj[key] ? 'true' : 'false');
                    }
                    else {
                        formData.append(key, obj[key]);
                    }
                }
            }
            return formData;
        };
        this.createPaymentChannel = async (request) => {
            return this.doPost(`${this.buildRequestURI(request.miniAppId, constant_1.Constant.PAYMENT_CHANNELS)}`, this.convertObjectToFormData(request));
        };
        this.updatePaymentChannel = async (request) => {
            return this.doPut(`${this.buildRequestURI(request.miniAppId, constant_1.Constant.PAYMENT_CHANNELS, request.channelId)}`, this.convertObjectToFormData(request));
        };
        this.buildRequestURI = (appId, path, ...pathVariables) => {
            let rs = `${this.version}${constant_1.Constant.APPS}`;
            if (!!appId) {
                rs += `/${appId}`;
            }
            if (!!path) {
                rs += path;
            }
            for (const pathVariable of pathVariables) {
                rs += `/${pathVariable}`;
            }
            return rs;
        };
        this.version = version;
        this.apiKeyHeaderName = apiKeyHeaderName;
        this.apiKey = apiKey;
        this.identityHeaderName = identityHeaderName;
        this.identity = identity;
        if (proxy) {
            this.proxy = proxy;
            this.isUseProxy = true;
        }
        this.headers = {
            [apiKeyHeaderName]: `Bearer ${this.apiKey}`,
            [identityHeaderName]: this.identity,
            'X-Sdk-Version': this.sdkVersion,
            'X-Sdk-Name': this.sdkName,
        };
    }
    setProxy(proxy) {
        this.proxy = proxy;
        this.isUseProxy = true;
    }
    cancelProxy() {
        this.proxy = null;
        this.isUseProxy = false;
    }
    async doGet(endpoint, params, options = {}) {
        var _a;
        try {
            this.validateInit();
            const response = await axios_1.default.get(`${this.domain}${endpoint}`, {
                params,
                headers: this.headers,
                proxy: this.isUseProxy && Object.assign({}, this.proxy),
            });
            return Object.assign({ error: response.data.err, message: response.data.msg }, response.data.data);
        }
        catch (error) {
            if (error instanceof axios_1.AxiosError) {
                return Object.assign({ error: (_a = (error.response ? error.response.status : error.status)) !== null && _a !== void 0 ? _a : -1, message: options.mapAxiosError
                        ? options.mapAxiosError(error)
                        : error.response
                            ? error.response.data.msg
                            : error.message }, {});
            }
            return Object.assign({ error: -1, message: error instanceof Error ? error.message : error }, {});
        }
    }
    async doPost(endpoint, params, options = {}) {
        var _a;
        try {
            this.validateInit();
            const response = await axios_1.default.post(`${this.domain}${endpoint}`, params, {
                headers: this.headers,
                proxy: this.isUseProxy && Object.assign({}, this.proxy),
            });
            return Object.assign({ error: response.data.err, message: response.data.msg }, response.data.data);
        }
        catch (error) {
            if (error instanceof axios_1.AxiosError) {
                return Object.assign({ error: (_a = (error.response ? error.response.status : error.status)) !== null && _a !== void 0 ? _a : -1, message: options.mapAxiosError
                        ? options.mapAxiosError(error)
                        : error.response
                            ? error.response.data.msg
                            : error.message }, {});
            }
            return Object.assign({ error: -1, message: error instanceof Error ? error.message : error }, {});
        }
    }
    async doPut(endpoint, params, options = {}) {
        var _a;
        try {
            this.validateInit();
            const response = await axios_1.default.put(`${this.domain}${endpoint}`, params, {
                headers: this.headers,
                proxy: this.isUseProxy && Object.assign({}, this.proxy),
            });
            return Object.assign({ error: response.data.err, message: response.data.msg }, response.data.data);
        }
        catch (error) {
            if (error instanceof axios_1.AxiosError) {
                return Object.assign({ error: (_a = (error.response ? error.response.status : error.status)) !== null && _a !== void 0 ? _a : -1, message: options.mapAxiosError
                        ? options.mapAxiosError(error)
                        : error.response
                            ? error.response.data.msg
                            : error.message }, {});
            }
            return Object.assign({ error: -1, message: error instanceof Error ? error.message : error }, {});
        }
    }
    async getMiniApps(appSlice) {
        return this.doGet(`${this.buildRequestURI()}`, Object.assign({}, appSlice));
    }
}
exports.BaseClient = BaseClient;
exports.default = { BaseClient };
