// zmp-openapi-nodejs/build/main/interface/payment/payment-channel.interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bank = exports.PaymentMethod = exports.Status = void 0;
var Status;
(function (Status) {
    Status[Status["ACTIVE"] = 0] = "ACTIVE";
    Status[Status["INACTIVE"] = 1] = "INACTIVE";
})(Status = exports.Status || (exports.Status = {}));
var PaymentMethod;
(function (PaymentMethod) {
    PaymentMethod[PaymentMethod["ZALOPAY"] = 0] = "ZALOPAY";
    PaymentMethod[PaymentMethod["MOMO"] = 1] = "MOMO";
    PaymentMethod[PaymentMethod["ATM"] = 2] = "ATM";
    PaymentMethod[PaymentMethod["CC"] = 3] = "CC";
    PaymentMethod[PaymentMethod["VNPAY"] = 4] = "VNPAY";
    PaymentMethod[PaymentMethod["COD"] = 5] = "COD";
    PaymentMethod[PaymentMethod["BANK"] = 6] = "BANK";
    PaymentMethod[PaymentMethod["PAYME_ATM"] = 7] = "PAYME_ATM";
    PaymentMethod[PaymentMethod["PAYME_CC"] = 8] = "PAYME_CC";
    PaymentMethod[PaymentMethod["ZALOPAY_SANDBOX"] = 9] = "ZALOPAY_SANDBOX";
    PaymentMethod[PaymentMethod["MOMO_SANDBOX"] = 10] = "MOMO_SANDBOX";
    PaymentMethod[PaymentMethod["ATM_SANDBOX"] = 11] = "ATM_SANDBOX";
    PaymentMethod[PaymentMethod["CC_SANDBOX"] = 12] = "CC_SANDBOX";
    PaymentMethod[PaymentMethod["VNPAY_SANDBOX"] = 13] = "VNPAY_SANDBOX";
    PaymentMethod[PaymentMethod["COD_SANDBOX"] = 14] = "COD_SANDBOX";
    PaymentMethod[PaymentMethod["BANK_SANDBOX"] = 15] = "BANK_SANDBOX";
    PaymentMethod[PaymentMethod["PAYME_ATM_SANDBOX"] = 16] = "PAYME_ATM_SANDBOX";
    PaymentMethod[PaymentMethod["PAYME_CC_SANDBOX"] = 17] = "PAYME_CC_SANDBOX";
    PaymentMethod[PaymentMethod["CUSTOM"] = 18] = "CUSTOM";
})(PaymentMethod = exports.PaymentMethod || (exports.PaymentMethod = {}));
var Bank;
(function (Bank) {
    Bank["ABBANK"] = "abbank";
    Bank["ACB"] = "acb";
    Bank["AGRIBANK"] = "agribank";
    Bank["BAC_A_BANK"] = "bacabank";
    Bank["BAOVIET_BANK"] = "baoviet";
    Bank["BIDC"] = "bidc";
    Bank["BIDV"] = "bidv";
    Bank["CAKE"] = "cake";
    Bank["CBBANK"] = "cbbank";
    Bank["CIMB"] = "cimb";
    Bank["COOP_BANK"] = "coopbank";
    Bank["DBS_BANK"] = "dbsbank";
    Bank["DONG_A_BANK"] = "dongabank";
    Bank["EXIMBANK"] = "eximbank";
    Bank["GPBANK"] = "gpbank";
    Bank["HDBANK"] = "hdbank";
    Bank["HONGLEONG_BANK"] = "hongleongbank";
    Bank["HSBC"] = "hsbc";
    Bank["IBK_HCM"] = "ibkhcm";
    Bank["IBK_HN"] = "ibkhn";
    Bank["INDOVINA_BANK"] = "indovinabank";
    Bank["KASIKORN_BANK"] = "kasikorn";
    Bank["KIENLONG_BANK"] = "kienlongbank";
    Bank["KOOKMIN_BANK_HCM"] = "kookminhcm";
    Bank["KOOKMIN_BANK_HN"] = "kookminhn";
    Bank["LIENVIETPOST_BANK"] = "lienvietpostbank";
    Bank["MBBANK"] = "mbbank";
    Bank["MSB"] = "msb";
    Bank["NAM_A_BANK"] = "namabank";
    Bank["NCB"] = "ncb";
    Bank["NONGHYUP_BANK_HN"] = "nonghyup";
    Bank["OCB"] = "ocb";
    Bank["OCEANBANK"] = "oceanbank";
    Bank["PGBANK"] = "pgbank";
    Bank["PUBLIC_BANK"] = "publicbank";
    Bank["PVCOM_BANK"] = "pvcombank";
    Bank["SACOMBANK"] = "sacombank";
    Bank["SAIGONBANK"] = "saigonbank";
    Bank["SCB"] = "scb";
    Bank["SEA_BANK"] = "seabank";
    Bank["SHB"] = "shb";
    Bank["SHINHAN_BANK"] = "shinhanbank";
    Bank["STANDARD_CHARTERED_BANK"] = "standardcharteredbank";
    Bank["TECHCOMBANK"] = "techcombank";
    Bank["TIMO"] = "timo";
    Bank["TPBANK"] = "tpbank";
    Bank["UBANK"] = "ubank";
    Bank["UNITED_OVERSEAS_BANK"] = "uob";
    Bank["VIB"] = "vib";
    Bank["VIET_A_BANK"] = "vietabank";
    Bank["VIET_BANK"] = "vietbank";
    Bank["VIETCAPITAL_BANK"] = "banviet";
    Bank["VIETCOMBANK"] = "vietcombank";
    Bank["VIETINBANK"] = "vietinbank";
    Bank["VPBANK"] = "vpbank";
    Bank["VRB"] = "vrb";
    Bank["WOORI_BANK"] = "wooribank";
})(Bank = exports.Bank || (exports.Bank = {}));
